stumpy/ostinato.py

```diff
@@ -0,0 +1,187 @@
+            h = j + 1
+        else:
+            h = 0
+
+        mp = stump(Ts[j], m, Ts[h], ignore_trivial=False)
+        si = np.argsort(mp[:, 0])
+        for q in si:
+            rad = mp[q, 0]
+            if rad >= bsf_rad:
+                break
+            for i in range(k):
+                if ~np.isin(i, [j, h]):
+                    QT = core.sliding_dot_product(Ts[j][q : q + m], Ts[i])
+                    rad = np.max(
+                        (
+                            rad,
+                            np.min(
+                                core._mass(
+                                    Ts[j][q : q + m],
+                                    Ts[i],
+                                    QT,
+                                    M_Ts[j][q],
+                                    Σ_Ts[j][q],
+                                    M_Ts[i],
+                                    Σ_Ts[i],
+                                )
+                            ),
+                        )
+                    )
+                    if rad >= bsf_rad:
+                        break
+            if rad < bsf_rad:
+                bsf_rad, ts_ind, ss_ind = rad, j, q
+
+    return bsf_rad, ts_ind, ss_ind
+
+
+def _get_central_motif(ts, rad, tsi, ssi, m):
+    """
+    Compare subsequences with the same radius and return the most central motif
+
+    Parameters
+    ----------
+    ts : list
+        List of time series for which to find the most central motif
+
+    rad : float
+        Best radius found by a consensus search algorithm
+
+    tsi : int
+        Index of time series in which `rad` was found first
+
+    ssi : int
+        Start index of subsequence in `tsi` that has radius `rad`
+
+    m : int
+        Window size
+
+    Returns
+    -------
+    rad : float
+        Radius of the most central consensus motif
+
+    tsi : int
+        Index of time series which contains the most central consensus motif
+
+    ssi : int
+        Start index of most central consensus motif within the time series `tsi`
+        that contains it
+    """
+    k = len(ts)
+    # Get nearest neighbors for ostinato hit
+    nn_ost, d_ost = _across_series_nearest_neighbors(ts, tsi, ssi, m)
+    # Alternative candidates with same radius
+    tsi_alt = np.flatnonzero(np.isclose(d_ost, rad))
+    ssi_alt = nn_ost[tsi_alt]
+    num_alt = len(tsi_alt)
+    d_alt = np.zeros((num_alt, k), dtype=float)
+    for i in range(num_alt):
+        _, d_alt[i] = _across_series_nearest_neighbors(ts, tsi_alt[i], ssi_alt[i], m)
+    d_alt_sum = np.sum(d_alt, axis=1)
+    if np.any(d_alt_sum < d_ost.sum()):
+        i_alt_best = np.argmin(d_alt_sum)
+        return rad, tsi_alt[i_alt_best], ssi_alt[i_alt_best]
+    else:
+        return rad, tsi, ssi
+
+
+def _across_series_nearest_neighbors(ts, qtsind, qssind, m):
+    k = len(ts)
+    q = ts[qtsind][qssind : qssind + m]
+    d = np.zeros(k, dtype=float)
+    nn = np.zeros(k, dtype=int)
+    nn[qtsind] = qssind
```

Comment:
Maybe I don't fully understand everything yet so this comment may seem uninformed. 

For safety, should we pre-emptively set `nn[qtsind] = -1` and `d[qtsind] = np.inf`?  Essentially, I am setting them to non-valid values to avoid a self match later.

Reply:
This is basically getting the minima from calls to `mass` on the different time series. I simplified it a bit to match `mass`'s  arguments.
The self-match is avoided in line 162, when comparing the distances with the radius.

Reply:
> This is basically getting the minima from calls to mass on the different time series. I simplified it a bit to match mass's arguments.

Yep, I understand this

> The self-match is avoided in line 162, when comparing the distances with the radius.

Okay, I think I get it now. So, while the distance for the self-match is zero, we are expecting that the radius will be greater than zero. However, I am concerned with edge cases where the radius may possibly be zero and so the self-match will be incorporated later. I understand that this is exceedingly rare but I tend to prefer programming "defensively" when there is little upfront cost to us (i.e., to instantiate the `d[qtsind] = np.inf`).

Reply:
I haven't thought through this but what happens when all of the time series have a subsequence that is the same in all of them? Wouldn't the radius for that subsequence be zero and the distance would also be zero?

Reply:
> Okay, I think I get it now. So, while the distance for the self-match is zero, we are expecting that the radius will be greater than zero. However, I am concerned with edge cases where the radius may possibly be zero and so the self-match will be incorporated later. I understand that this is exceedingly rare but I tend to prefer programming "defensively" when there is little upfront cost to us (i.e., to instantiate the d[qtsind] = np.inf).

In the current implementation the no time series is skipped, so instantiating `d[qtsind]` shouldn't be necessary.

> I haven't thought through this but what happens when all of the time series have a subsequence that is the same in all of them? Wouldn't the radius for that subsequence be zero and the distance would also be zero?

Yes. All the matching subsequences would be at the same point. In that case all these subsequences are the most central one. I'll add a default behaviour for this edge case.